Model/Api/OrderManagement.php

```diff
@@ -205,34 +232,39 @@ private function saveUpdateOrder($reference, $type, $display_id, $storeId)
                 __('Missing required parameters.')
             );
         }
-        if ($type === 'rejected_irreversible' && $this->orderHelper->tryDeclinedPaymentCancelation($display_id)) {
-            $this->response->setHttpResponseCode(200);
-            $this->response->setBody(json_encode([
-                'status' => 'success',
-                'message' => 'Order was canceled due to declined payment: ' . $display_id,
-            ]));
-        } elseif ($type === 'failed_payment') {
+        if ($type === 'failed_payment') {
             $this->orderHelper->deleteOrderByIncrementId($display_id);
 
             $this->response->setHttpResponseCode(200);
             $this->response->setBody(json_encode([
                 'status' => 'success',
                 'message' => 'Order was deleted: ' . $display_id,
             ]));
-        } else {
-            $this->orderHelper->saveUpdateOrder(
-                $reference,
-                $storeId,
-                $this->request->getHeader(ConfigHelper::BOLT_TRACE_ID_HEADER),
-                $type
-            );
+            return;
+        }
+
+        $this->handleCheckboxes($display_id);
```

Comment:
Perhaps we should handle checkboxes somewhere else, not in every hook. This way we can optimize things a bit by eliminating addCommentToStatusHistoryIfNotExists method and it's logic (that I am not very fond of).  \Bolt\Boltpay\Helper\Order::dispatchPostCheckoutEvents seems like a good candidate, it is run only once after order is approved.
@YevhenShBolt @matt-thomason @njfaries

Reply:
@var-bolt I like @danac-gs proposal, but actually it means that we will look into checkboxes only in first, 'pending' hook.
If for some reason 'pending' hook doesn't contain checkboxes, but the next hook ('completed' ) contains we don't handle checkboxes.
Do you think it's a real issue or not?

Reply:
I think it should be okay! But I will check with folks and get back on this comment

Reply:
@vitaliyreznikov @matt-thomason @YevhenShBolt @njfaries 
 dispatchPostCheckoutEvents() is called at two places, in Controller\ReceivedUrlTrait::execute and Helper\Order::saveUpdateOrder and the full transaction info is not available in former one, so it is not a good canditate, you're right. There's this place just before Helper\Order::saveUpdateOrder:
```php
if ($order->getState() === OrderModel::STATE_PENDING_PAYMENT) {
    $this->resetOrderState($order);
}
```
This condition passes only once when order is approved and status changes from our pending_payment to magento default initial state - new / pending. We have all the data here ($transaction, $order) so we can pass it along to checkboxes handler. The handler gets all the data it needs, $transaction->checkboxes and $order, should just loop through checkboxes and either call specific handler for everyone of them (my proposal) or do it all in place (your solution). Moreover, $order->addStatusHistoryComment can be called there to record checkboxes, without the need of addCommentToStatusHistoryIfNotExists logic. As it is called one time only in Helper\Order::resetOrderState. Please consider this, it is much simpler than proposed solution. In a nutshell:
1. find a place in transaction lifetime (hooks) that executes only once after the order is approved (both bolt transaction data and magento order should be available here)
2. inject your method call there. It should be defined in an external class, Helper/CheckboxesHandler you already created.

Reply:
Yeah, I didn't realize that I can find checkboxes data as `$transaction->checkboxes` Now I got it, thank you for detail explanation.

Reply:
@danac-gs This is an example of `$transaction->checkboxes`
```
Array (    [0] => stdClass Object  (
            [external_id] => test_checkbox
	    [label] => test checkox
            [default_value] => 1
            [position] => 0
        )
)
```

It doesn't contain field **category** (like NEWSLETTER) so we should use data from webhook and we don't have it in **Helper\Order::saveUpdateOrder** right now. What do you think?

Reply:
@vitaliyreznikov chances are external_id == category. If not this should be raised and probably fixed on another level, BX / checkout / core team, not sure. In my opinion, and from my experience all the relevant (checkboxes and other) data should be held in fetch transaction response. Lets at least try to fix the design before complicating the code on plugin level.
cc @var-bolt @matt-thomason @njfaries @ethanwayda22

Reply:
@daisy1754 @nirajjayantbolt thoughts ^^^

Reply:
Let's wait for more opinions, but from my point of view, although we found a sort of issue that should be fixed on the server side, it's still better to use primary information (webhook data) than the second one (data from the transaction).
It seems like a more scalable approach: if we need down the road to add additional checkboxes type or something else it's more simple doing webhook data.

If we want to avoid checking if comment exists we can pass all webhook data into `Helper\Order::saveUpdateOrder` It's not a very big deal and can be helpful in other situation in the future.

Reply:
@danac-gs @vitaliyreznikov @daisy1754 From the tech spec, it seems we were operating under the assumption that category would be explicitly provided. Here's a sample from the M2 spec:
```
"checkboxes": [
    {
      "checkbox_id": "subscription",
      "text": "Sign me up for the newsletter!",
      "value": true,
      "category": "newsletter",
      "features": [
        "subscribe_to_platform_newsletter"
      ]
    }
  ]
```

@vitaliyreznikov I also agree with you that relying on the webhook data is a stronger solution than the transaction data - Bolt should ideally be the source of truth since checkbox configs are coming from Bolt anyways, and it allows us more flexibility in the future.

@daisy1754 Can you clarify re: the category?

Reply:
@matt-thomason @vitaliyreznikov @daisy1754 
So far the source of truth has always been the transaction data, ie. we always fetch the transaction info when get the hook because hook has not all the info needed. At this point we have already fetched the transaction. The question is, since there's a checkboxes section in transaction why it differs from the hook, is that intentional (why?) or an (easy to fix?) oversight. Anyway, I agree with this approach
> If we want to avoid checking if comment exists we can pass all webhook data into Helper\Order::saveUpdateOrder It's not a very big deal and can be helpful in other situation in the future.

Better pas one more variable and insert just one external call at the appropriate place then change / complicate already complex logic. In general, we should try to avoid altering the things that work / doing it as little as possible, unless the change is driven by code simplification or refactoring. Whenever we are about to add more complexity we should look back and check if there's another way. At least that's how I see it.

Reply:
@var-bolt thought on why GET transaction response and payload in webhook is different?

Reply:
re: wether we should rely on webhook or get transaction, I don't have a strong preference but I agree that I prefer to stick to one data source. I think once we complete refactoring of webhook [1] we can fully switch to webhook but I feel like that would be medium size project 
 [1] https://docs.google.com/document/d/159lDLJHjBPOCqLqPdLbabriS-OXEOX9lTIFcbsTOulY/edit?ts=5e33437c#

Reply:
okay so couple of updates from convo in slack + me reading hail code:
1) difference between webhook and GET transaction is intended (I am personally not sure if I agree with this, but it's a longer discussion)
2) we'll refactor webhook payload and webhook will contain transaction view but that project is a few month away from productionize
3) external ID != category. if we want to access category we'd have to rely on webhook now

Given all + product's desire to include this in upcoming M2 release, I support relying on custom payload in hook here. we can re-discuss when (2) is ready to integrate.

@matt-thomason @danac-gs @vitaliyreznikov

Reply:
@daisy1754 <3

Let's get this merged! Should we do 1 (or) 2 we can revisit this at a later point in time

Reply:
@danac-gs about your comment in this thread from Feb 7th
```
if ($order->getState() === OrderModel::STATE_PENDING_PAYMENT) {
    $this->resetOrderState($order);
}
```
I've checked and at least on my test env. `$order->getState()` is `OrderModel::STATE_NEW` when we handle `pending ` hook and `OrderModel::STATE_PROCESSING` when we handle `payment `hook. So code inside if is never run.

Could you please check if we have any issues here?

Reply:
@vitaliyreznikov the order is created in state / status pending_payment / pending_payment. \Bolt\Boltpay\Helper\Order::resetOrderState is the place where it's moved into new / pending.
However I just realized in pre-auth flow resetOrderState is called earlier, in success redirect call, \Bolt\Boltpay\Controller\ReceivedUrlTrait::execute. So the call from hook is skipped. Chances are that call is not needed and can be removed from there. I will take a look into it.

Reply:
@vitaliyreznikov based on a quick test we can remove it from ReceivedUrlTrait, and rely on it being called in hooks. Will run more tests, but you can check as well if you have the time,

Reply:
@danac-gs I also think your proposal should work, but if we implement it and remove **resetOrderState** from **ReceivedUrlTrait** it means that we set "new / pending" later that now and order will be sent to ERP system later. Do we need this change?
Anyway, this refactoring looks too big to be included in this PR.

If our goal is to handle checkboxes only once and do it in `Helper\Order::saveUpdateOrder` we can rely on bolt hook type (parameter $hookType) and handle checkboxes only if $hookType=='pending'.
And we can refactor it later, in separate PR. What do you think?

Reply:
@vitaliyreznikov calling **resetOrderState** from hook is OK. It already works that way for non pre-auth orders. On the other hand, pending hook can potentially fail and we end up unnecessary doing this in loop. I'll be running tests over the weekend with resetOrderState removed from ReceivedUrlTrait and if all goes well create a separate pulls request. Early Monday morning our time at latest.

Reply:
@danac-gs Ok, sounds good.

Reply:
@vitaliyreznikov I tested this ^^ and it works. Will create a PR for it soon.

Reply:
@vitaliyreznikov https://github.com/BoltApp/bolt-magento2/pull/554